spec/index.adoc

```diff
@@ -629,17 +632,52 @@ In order to make a query call to canister, the user makes a POST request to `/ap
 * `method_name` (`text`): Name of the canister method to call
 * `arg` (`blob`): Argument to pass to the canister method
 
+The HTTP response to a query call can contain a list of signatures for the returned response produced by the individual IC nodes that computed the same returned response.
+Every such signature (whose type is denoted as `node_signature_of_http_query_response`) is a CBOR (see <<cbor>>) map with the following fields:
+
+* `timestamp` (`nat`): the timestamp of the signature.
+* `signature` (`blob`): the actual signature.
+* `identity` (`principal`): the principal of the node producing the signature.
+
 If the call resulted in a reply, the response is a CBOR (see <<cbor>>) map with the following fields:
 
-* `status` (`text`): `replied`
-* `reply`: a CBOR map with the field `arg` (`blob`) which contains the reply data.
+* `status` (`text`): `"replied"`
+* `reply`: a CBOR map with the field `arg` (`blob`) which contains the reply data `<R>`.
+* `node_signatures` (`[* node_signature_of_http_query_response]`): a list of node signatures for the returned query response
+  computed for the following value
++
+....
+hash_of_map({
+  status: "replied",
+  reply: {arg: <R>},
+  timestamp: <t>,
+  request_id: hash_of_map({request_type: "query", sender: <sender>, nonce: <nonce>, ingress_expiry: <ingress_expiry>, canister_id: <canister_id>, method_name: <method_name>, arg: <arg>})
+})
+....
++
+where `hash_of_map` is the <<hash-of-map, representation-independent hash>> and `<t>` is the timestamp of the node when the signature was produced.
 
 If the call resulted in a reject, the response is a CBOR map with the following fields:
 
-* `status` (`text`): `rejected`
+* `status` (`text`): `"rejected"`
 * `reject_code` (`nat`): The reject code (see <<reject-codes>>).
 * `reject_message` (`text`): a textual diagnostic message.
-* `error_code` (text): an optional implementation-specific textual error code (see <<error-codes>>).
+* `error_code` (`text`): an optional implementation-specific textual error code (see <<error-codes>>).
+* `node_signatures` (`[* node_signature_of_http_query_response]`): a list of node signatures for the returned query response
+  computed for the following value
++
+....
+hash_of_map({
+  status: "rejected",
+  reject_code: <reject_code>,
+  reject_message: <reject_message>,
+  error_code: <error_code>,
+  timestamp: <t>,
+  request_id: hash_of_map({request_type: "query", sender: <sender>, nonce: <nonce>, ingress_expiry: <ingress_expiry>, canister_id: <canister_id>, method_name: <method_name>, arg: <arg>})
+})
+....
++
+where `hash_of_map` is the <<hash-of-map, representation-independent hash>> and `<t>` is the timestamp of the node when the signature was produced.
```

Comment:
How does the client know the timestamp used by the replica? Should this timestamp be added to the response?

Also, is the request_id here the same that's used by the client to sign the request?
If yes, maybe simply state this, instead of (redundantly) repeating it's calculation algorithm?
If no (different fields maybe), better don't use the same name here.

Reply:
> How does the client know the timestamp used by the replica? Should this timestamp be added to the response?

The replica's timestamp is part of a node signature record together with the node principal and the actual signature (please see above).

> Also, is the request_id here the same that's used by the client to sign the request?

Yes.

> If yes, maybe simply state this, instead of (redundantly) repeating it's calculation algorithm?

I've replaced `hash_of_map({request_type: "query", sender: <sender>, nonce: <nonce>, ingress_expiry: <ingress_expiry>, canister_id: <canister_id>, method_name: <method_name>, arg: <arg>})` by `hash_of_map(content)`.

Reply:
Ah, with the CDDL I see it now. Thanks!